extensions-contrib/sqlserver-metadata-storage/src/main/java/io/druid/metadata/storage/sqlserver/SQLServerConnector.java

```diff
@@ -0,0 +1,183 @@
+    log.info("Configured Sql Server as metadata storage");
+  }
+
+  public static class CustomStatementRewriter implements StatementRewriter
+  {
+      @Override
+    public RewrittenStatement rewrite(String sql, Binding params, StatementContext ctx)
+    {
+
+      String sqlUpdated = sql
+        .replaceAll( "(?i)BOOLEAN NOT NULL DEFAULT FALSE", "BIT NOT NULL DEFAULT (0)" )
+        .replaceAll( "(?i)BOOLEAN NOT NULL DEFAULT TRUE", "BIT NOT NULL DEFAULT (1)" )
+        .replaceAll( "(?i)BOOLEAN DEFAULT FALSE", "BIT NOT NULL DEFAULT (0)" )
+        .replaceAll( "(?i)BOOLEAN DEFAULT TRUE", "BIT NOT NULL DEFAULT (1)" )
+        .replaceAll( "(?i)BOOLEAN", "BIT" )
+        .replaceAll( "(?i)TRUE", "1" )
+        .replaceAll( "(?i)FALSE", "0" )
+        ;
+      return ( new ColonPrefixNamedParamStatementRewriter() ).rewrite( sqlUpdated, params, ctx );
+    }
+  }
+
+  @Override
+  protected String getPayloadType() {
+    return PAYLOAD_TYPE;
+  }
+
+  @Override
+  protected String getSerialType()
+  {
+    return SERIAL_TYPE;
+  }
+
+  @Override
+  protected int getStreamingFetchSize()
+  {
+    return DEFAULT_STREAMING_RESULT_SIZE;
+  }
+
+  @Override
+  public boolean tableExists(final Handle handle, final String tableName)
+  {
+    return !handle.createQuery("SELECT * FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_NAME = :tableName")
+      .bind("tableName", tableName)
+      .map(StringMapper.FIRST)
+      .list()
+      .isEmpty();
+  }
+
+  /**
+   *
+   * @see "http://stackoverflow.com/questions/1197733/does-sql-server-offer-anything-like-mysqls-on-duplicate-key-update"
+   *
+   * {@inheritDoc}
+   */
+  @Override
+  public Void insertOrUpdate(
+      final String tableName,
+      final String keyColumn,
+      final String valueColumn,
+      final String key,
+      final byte[] value
+  ) throws Exception
+  {
+    return getDBI().withHandle(
+      new HandleCallback<Void>()
+      {
+        @Override
+        public Void withHandle(Handle handle) throws Exception
+        {
+          handle.createStatement( String.format(
+              "MERGE INTO %1$s WITH (UPDLOCK, HOLDLOCK) as target"
+              + " USING "
+              + " (:key, :value) as source (%2$s, %3$s)"
+              + " ON"
+              + " (target.%2$s = source.%2$s)"
+              + " WHEN MATCHED THEN UPDATE SET %3$s = :value"
+              + " WHEN NOT MATCHED THEN INSERT (%2$s, %3$s) VALUES (:key, :value)",
+              tableName,
+              keyColumn,
+              valueColumn )
+            )
+            .bind("key", key)
+            .bind("value", value)
+            .execute();
+
+          return null;
+        }
+      }
+    );
+  }
+
+    @Override
+    public DBI getDBI()
+    {
+      return dbi;
+    }
+
+  /**
+   *
+   * @see java.sql.SQLException#getSQLState()
+   * @see "https://github.com/spring-projects/spring-framework/blob/master/spring-jdbc/src/main/java/org/springframework/jdbc/support/SQLStateSQLExceptionTranslator.java"
```

Comment:
In the future, since this references master, this file may or may not exist. If you link to the file in a git tag instead of in master, it will persist if the file is moved or deleted.